### ISSUE TITLE
Something wrong in handling list

### ISSUE THREAD

**Issue by smartepsh** [corka149/jsonpatch#8]
**Describe the bug**

1. The array is a legal json structured type, but can't apply a patch for a list directly. (Maybe it's your feature...)
2. Can't handle list well. (see examples below)
3. (Suggestion) Swap arguments of `Jsonpatch.apply_patch/2` so that the pipeline can be used. `map |> Jsonpatch.apply_patch!(diff_1) |> Jsonpatch.apply_patch!(diff_2)`

**To Reproduce**

Example 1:

```elixir
  base = %{list: [1, 2, 3]}
  target = %{list: [2, 3, 1]}
  diff = Jsonpatch.diff(base, target)
  Jsonpatch.apply_patch!(diff, base)
```

Example 2:

```elixir
  base = %{list: [%{a: 1}, %{b: 2}, %{c: 3}]}
  target = %{list: [%{b: 2}, %{c: 3}, %{a: 1}]}
  diff = Jsonpatch.diff(base, target)
  Jsonpatch.apply_patch!(diff, base)
```
**Expected behavior**

I hope the result of `Jsonpatch.after_patch!(diff, base)` is equal to `target` .

**Screenshots**

Example 1:

<img width="1340" alt="image" src="https://user-images.githubusercontent.com/3273295/167575287-30157417-8413-4e9a-96d5-917c439dbb88.png">

Example 2:

<img width="823" alt="image" src="https://user-images.githubusercontent.com/3273295/167575453-32171b5c-877f-4945-a1a4-a21a6b7720ab.png">

**Desktop (please complete the following information):**
- Elixir - 1.13.3
- Erlang - 24.2.1
- System - MacOS 12.3.1

**Comment by corka149** [comment]
Thanks for the report!

> (Maybe it's your feature...)

Indeed. This is my fault that there is no hint in the documentation, yet. The issue is the map and not the list because it is atom based map. Atoms as keys are not supported, yet. 

Why? So far, I saw only JSON parser that create maps with strings as keys because they are safer in case not trusted JSON is parsed. Else it would be possible to trigger the creation of atoms from the outside (when we are talking about REST interfaces for example). I followed this philosophy but never stated it anywhere.
(The path of the patch operation must be converted into an atom.)

But today I think it should be possible to use something like an option. Example:
```Elixir
Jsonpatch.apply_patch!(diff, base, allow_atoms: true)
```

In addition the error message could be more useful.

- [ ] Improve error message in case of maps with atom keys
- [x] Mention danger of using maps with atom keys
- [x] Add option to allow the creation of atoms

I have no spare time this week. If its urgent or you are interest in solving it - feel free to create a PR. :pray:

**Comment by smartepsh** [comment]
Thanks for the reply. 

I hadn't noticed before that this was a problem with atom keys 😮‍💨 I have another example about the list with the string key's map.

### Example 3

```elixir
  base = %{"key" => []}
  target = %{"key" => [1, 2, 3]}
  diff = Jsonpatch.diff(base, target)
  Jsonpatch.apply_patch!(diff, base)
```

<img width="606" alt="image" src="https://user-images.githubusercontent.com/3273295/167673985-3975bed5-5ef5-49d8-af87-73234cc763d9.png">

It restores only one element of the list. I think it's about the diff order.

**Comment by corka149** [comment]
You are right. `Jsonpatch.apply_patch` should follow the order.

On the other hand, the result of the diff should be in the same orders as they are detected. That should be changed.

- [x] Produce diff in the same order as they are detected

**Comment by smartepsh** [comment]
> You are right. `Jsonpatch.apply_patch` should follow the order.
> 
> On the other hand, the result of the diff should be in the same orders as they are detected. That should be changed.
> 
> * [x]  Produce diff in the same order as they are detected

This is more important to us, so I took care of that first. Please. https://github.com/corka149/jsonpatch/pull/9

**Comment by corka149** [comment]
Nice - thanks for the contribution! I merged it and published version 0.12.1. :-)

**Comment by corka149** [comment]
The progress looks good so far:
```
iex(1)> base = %{list: [1, 2, 3]}
%{list: [1, 2, 3]}
iex(2)> target = %{list: [2, 3, 1]}
%{list: [2, 3, 1]}
iex(3)> diff = Jsonpatch.diff(base, target)
[
  %Jsonpatch.Operation.Replace{path: "/list/0", value: 2},
  %Jsonpatch.Operation.Replace{path: "/list/1", value: 3},
  %Jsonpatch.Operation.Replace{path: "/list/2", value: 1}
]
iex(4)> Jsonpatch.apply_patch!(diff, base, keys: :atoms)
%{list: [2, 3, 1]}
```

I think only tests are missing.

**Comment by corka149** [comment]
The change is now in main and released under version 0.13.0.
